keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -11,6 +11,7 @@
 - [Proposal](#proposal)
   - [API Changes](#api-changes)
     - [Subresource](#subresource)
+    - [Validation](#validation)
```

Comment:
I don't have good place to put that comment, so let me put it here.

I wanted to get back to the aspect of interactions with scheduler. Currently the KEP is saying:
```
Scheduler may, in parallel, assign a new Pod to the Node because it uses cached Pods to compute Node allocable values. If this race condition occurs, Kubelet resolves it by rejecting that new Pod if the Node has no room after Pod resize.

Note: After a Pod is rejected, the scheduler could try to reschedule the replacement pod on the same node that just rejected it. This is a general statement about Kubernetes and is outside the scope of this KEP.
```

I think that was fine when the KEP was created in 2020/2021 timeframe, but I'm no longer convinced this is good enough now.

We're already in the world where people are running a bunch of tightly-coupled workload on k8s (AI/ML Training being one example here). We don't have a perfect solution for gang-scheduling yet, but things like Kueue make it work reasonable and there are multiple people (including myself) thinking now how to make it actually reliable.

In-Place updates has a big potential to break that visibly more. And this is risky, because these jobs are often relatively large and running on extremely expensive software, so the need for recreation of the pod, going through the whole cycle again etc. can have pretty non-negligible cost impact.


I don't have a reasonable alternative (ensuring that kube-scheduler acknowledges the change before Kubelet can consume it sounds like an overkill to me), but I believe we should get back to this problem and rethink it again.

@tallclair

Reply:
> In-Place updates has a big potential to break that visibly more. And this is risky, because these jobs are often relatively large and running on extremely expensive software, so the need for recreation of the pod, going through the whole cycle again etc. can have pretty non-negligible cost impact.
> 
> I don't have a reasonable alternative (ensuring that kube-scheduler acknowledges the change before Kubelet can consume it sounds like an overkill to me), but I believe we should get back to this problem and rethink it again.
> 
> @tallclair

This section refers to resolving races by rejecting a "overbooked" new pod arriving at the Kubelet competing with a resize that got admitted first. Such a pod is not yet running, and recreation flow in this example would be a job controller tossing the failed pod out of etcd and creating a new pod object.. and I think that part can be optimized with scheduler and controller change, shouldn't be too hard. Already running pods would not be affected by a resize request, other than them not getting the desired resources due to availability. Is there a concrete example the help understand this scenario better?

The design of bringing scheduler into the workflow of approving pod resize was an idea that we initially explored but abandoned due to the that state-machine fragility that it introduced. And it still would not solve the multiple-scheduler race condition problem.

Reply:
I agree with @vinaykul, and I still think resolving this should be outside the scope of this KEP, but I do think it's worth revisiting this problem.

BTW, this is an idea I was discussing with @SergeyKanzhelev recently, which I think could be a potential solution to this problem: https://docs.google.com/document/d/1O9rDeYhGWMbtg-DQmhqKvNsQSHWJSwvSEeNpXTKMvZg/edit

Reply:
> Already running pods would not be affected by a resize request, other than them not getting the desired resources due to availability. Is there a concrete example the help understand this scenario better?

I agree that already running pods would not be affected. The problem is tightly coupled workloads.
Usecase: my AI/ML training job consists of 1000 pods and I need all of them to be up and running to make progress. In other words - if any of those pods is not running, the rest of them is also useless (and effectively wasting resources).


> The design of bringing scheduler into the workflow of approving pod resize was an idea that we initially explored but abandoned due to the that state-machine fragility that it introduced. And it still would not solve the multiple-scheduler race condition problem.

I don't think it's a valid argument. We need to ensure that we're solving real problems - the argument that something would be complex so we won't do that is not an answer.

> BTW, this is an idea I was discussing with @SergeyKanzhelev recently, which I think could be a potential solution to this problem: https://docs.google.com/document/d/1O9rDeYhGWMbtg-DQmhqKvNsQSHWJSwvSEeNpXTKMvZg/edit

Thanks for sharing that. I need to think about it more (I don't think scalability is a risk here, though the reliability implications would need to be carefully thought through), but indeed it seems like a valid path forward.
I personally don't think it should be out-of-scope for this KEP, but if I'm outvoted, can we at the very least explicitly say that:
(a) solving this problem is out-of-scope for this KEP
(b) we should solve it soon-after
(c) here is the idea that we will explore to solve it
?

Reply:
@thockin - I 100% agree that "in-place resiz is an important problem".
What you described in the first paragraph is what I was envisioning initially, and that actually doesn't sound super additive to me - I mean we can technically do that later, but would require significant changes, which means that maybe we should actually do them from the beginning.

That said, the doc that @tallclair linked above (the brainstom one) actually sounds much more additive to me. It has some drawbacks, but that shows that there are potential paths forwards that are clearly additive.
That's why I could live with the "but if I'm outvoted" part from my comment above.

But I would like to see that prioritized and ensure that we will do that - so maybe that means making it a GA criteria?

Reply:
What would you propose as GA criteria?  Something like "An approved design for scheduling reservations which prevents races between multi-pod scheduling and resizes" ?

Reply:
@wojtek-t Having scheduler approve resize requests was my first naive approach ( https://sched.co/MPcJ ). We subsequently this in detail during node meetings and reached a consensus that once scheduler has found a home for an unbound pod, its role ends. And all subsequent interactions such as resize is business between the API and node. This approach provided simplicity and a clean separation of concerns - avoiding complexity and fragility for little gain.

The scheduler hears the same music as the node, and avoids scheduling new pods to the node by using max(requests, allocatedResources), but yes there is a small (guessing) chance of a race, which as I mentioned, can be mitigated by rescheduling pods impacted by losing a race.

@dchen1107 had some great suggestions about other ways scheduler or another controller could assist e.g evicting lower priority pods for aiding resize (or PodReservation in @tallclair's doc for your use case) can be explored further as add-ons. Those improvements those were intentionally left out of scope because this is a large, complex KEP .. I feel we should avoid feature-creep, and work towards getting the simpler design into rock-solid shape first.

Reply:
> What would you propose as GA criteria? Something like "An approved design for scheduling reservations which prevents races between multi-pod scheduling and resizes" ?

Yes - something like that.
I like Tim's brainstorming doc, so maybe it's a matter for figuring out the details and getting consensus for it?
But I'm happy with other approaches too as long as we will eventually address that.

Reply:
Forgive me if I'm missed that in this KEP but I think we need:
- a callout on it in Risks/Mitigations section (describe the risk, say that we will explore the mitigation and maybe link @tallclair 's doc as one potential path forward?)
- [ideally, but I'm not sure you agree with that] have that as GA criteria too?

Reply:
I disagreed to document this as a risk/mitigation in the KEP or make it a GA criteria. Instead, framing it as a non-goal with a detailed explanation of the scheduler limitations and potential future enhancements is a more accurate and constructive approach IMHO. 

This is because the issue is not unique to in-place resizing; it stems from inherent limitations in the scheduler's ability to maintain a perfectly synchronized view of resource availability in a dynamic environment. These limitations exist today and can affect any pod, regardless of whether it's being resized. Both Vinay and called out the scheduler issue when the project was first initiated at SIG Scheduler and SIG Architecture, but didn't receive much supports back then. Again those are not new issues.  

There are several attempts to address this kind of issue, and I also proposed something at: https://shorturl.at/PnzW8

Reply:
By the way, this proposal from SIG Storage  https://github.com/kubernetes/enhancements/pull/4875  which might introduce the race condition for scheduler too.

Reply:
> This is because the issue is not unique to in-place resizing;

This is true, but before this proposal this was affecting us only in two main scenarios:
(a) setup with multiple schedulers (which is very rare, and users doing that definitely understand the risks)
(b) decreasing allocatable resources on nodes [which generally isn't something that's happening frequently]

[There are other situations like pod being deleted and not yet observed by scheduler, but that's not something that I'm super concerned, because it won't result in scheduling a pod that later won't fit].

This proposal will make it waaaay wider problem, because now many more pods will be upsizing (e.g. via VPA). So while I agree it's not a new problem, we will make it much more frequent. And we shouldn't try to hide it.

I buy Tim's arguments that it shouldn't block this proposal for the sake of making progress, but let's call it out as a risk.

Reply:
> a callout on it in Risks/Mitigations section

It's already listed in the failure modes section, but I added an explicit call out in the risks section.

I've also added it as a non-goal. Even if we decide it's a blocker for GA, it's a large enough design that the solution will need a separate KEP.

Reply:
sgtm

Reply:
One more point: There is an ongoing effort for resizable node, which amplify the scheduler limitations discussed here dramatically.

Reply:
Thanks Tim!

And yes - I agree that resizable nodes would make it much more problematic, it's something I'm actively thinking about...

Reply:
@dchen1107 and @wojtek-t can you please share a link to resizable node effort mentioned. 

Thanks @tallclair